File: get_stars.py
import requests
from bs4 import BeautifulSoup

def _count_stars_from_url(url: str) -> int:
  response = requests.get(url)
  soup = BeautifulSoup(response.text, 'html.parser')
  star_counts = [
    int(anchor.text.strip())
    for anchor in soup.select('a.muted-link.mr-3')
    if 'stargazers' in anchor.attrs['href']
  ]
  print(star_counts)
  next_button = soup.select_one('a.btn.btn-outline.BtnGroup-item:last-child')
  if next_button and 'Next' in next_button.text:
    next_url = next_button['href']
    return sum(star_counts) + _count_stars_from_url(next_url)
  return sum(star_counts)

def get_stars(user_id: str) -> int:
  return _count_stars_from_url(f'https://github.com/{user_id}?tab=repositories')

if __name__ == '__main__':
  print(get_stars('junhoyeo'))

File: actions/parseUser/getStars.ts
import axios from 'axios';
import cheerio from 'cheerio';

import getSumOfNumberArray from '../../utils/getSumOfNumberArray';
import { getTotalStarsForUser, hasGitHubToken, getGitHubToken } from '../../utils/githubGraphQL';

function parseStarCountText(text: string): number {
  const normalized = text.replace(/,+/g, '').trim().toLowerCase();
  const match = normalized.match(/^([\d.]+)([km])?$/);
  if (!match) {
    return Number(normalized) || 0;
  }
  const value = parseFloat(match[1]);
  if (!Number.isFinite(value)) {
    return 0;
  }
  if (match[2] === 'k') {
    return Math.round(value * 1000);
  }
  if (match[2] === 'm') {
    return Math.round(value * 1000000);
  }
  return value;
}

async function _countStarsFromURL(url: string): Promise<number> {
  const { data: html } = await axios.get(url);
  const document = cheerio.load(html);
  const starCounts = document('a[href*="stargazers"]')
    .toArray()
    .flatMap((anchorReference) => {
      const anchor = cheerio(anchorReference);
      const href = anchor.attr('href');
      if (href && href.includes('stargazers')) {
        return [
          parseStarCountText(anchor.text())
        ];
      }
      return [];
    });
  const currentCounts = getSumOfNumberArray(starCounts);

  const nextButton = document('a.next_page').first();
  if (nextButton?.text().includes('Next')) {
    let nextURL = nextButton.attr('href') ?? '';
    if (!nextURL?.startsWith('http')) {
      nextURL = `https://github.com` + nextURL
    }
    if (nextURL) {
      const nextCounts = await _countStarsFromURL(nextURL);
      return currentCounts + nextCounts;
    }
  }
  return currentCounts;
}

export default async function getStars(userID: string): Promise<number> {
  if (hasGitHubToken()) {
    try {
      const token = getGitHubToken();
      return await getTotalStarsForUser(token, userID);
    } catch (error) {
      console.warn(`[getStars] GraphQL failed for ${userID}, falling back to web scraping:`, error);
    }
  }

  return await _countStarsFromURL(`https://github.com/${userID}?tab=repositories`);
}
